File: real_pir_protocol.cpp
#include <algorithm>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <random>
#include <string>
#include <thread>
#include <vector>

#ifdef _WIN32
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <windows.h>
#endif

namespace fs = std::filesystem;

static std::string nowMs() {
    using namespace std::chrono;
    auto ms = duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
    return std::to_string(ms);
}

static double secsSince(const std::chrono::steady_clock::time_point &start) {
    using namespace std::chrono;
    return duration_cast<duration<double>>(steady_clock::now() - start).count();
}

static void printDivider() {
    std::cout << std::string(50, '=') << "\n";
}

// Read full text file consisting of '0' and '1' chars into vector<int> bits
static bool readBitsFile(const fs::path &path, std::vector<int> &outBits) {
    std::ifstream in(path, std::ios::in | std::ios::binary);
    if (!in.is_open()) return false;
    std::string s;
    in.seekg(0, std::ios::end);
    s.resize(static_cast<size_t>(in.tellg()));
    in.seekg(0, std::ios::beg);
    in.read(&s[0], static_cast<std::streamsize>(s.size()));
    outBits.reserve(s.size());
    for (char c : s) {
        if (c == '0') outBits.push_back(0);
        else if (c == '1') outBits.push_back(1);
    }
    return true;
}

// Write bits as '0'/'1' chars in chunks
static bool writeBitsFile(const fs::path &path, const std::vector<int> &bits) {
    std::ofstream out(path, std::ios::out | std::ios::binary | std::ios::trunc);
    if (!out.is_open()) return false;
    const size_t chunkSize = 1000000; // 1M bits at a time
    std::string buffer;
    buffer.reserve(chunkSize);
    size_t i = 0;
    while (i < bits.size()) {
        buffer.clear();
        const size_t end = std::min(bits.size(), i + chunkSize);
        for (; i < end; ++i) buffer.push_back(bits[i] ? '1' : '0');
        out.write(buffer.data(), static_cast<std::streamsize>(buffer.size()));
    }
    return true;
}

// Convert vector<int> bits to bytes and write to binary file in chunks
static bool writeBitsAsBinaryVideo(const fs::path &outPath, const std::vector<int> &bits) {
    std::ofstream out(outPath, std::ios::out | std::ios::binary | std::ios::trunc);
    if (!out.is_open()) return false;
    const size_t chunkBits = 1000000; // process 1M bits per chunk
    size_t i = 0;
    std::vector<unsigned char> byteBuf;
    byteBuf.reserve(chunkBits / 8 + 8);
    while (i < bits.size()) {
        const size_t end = std::min(bits.size(), i + chunkBits);
        byteBuf.clear();
        for (; i < end; i += 8) {
            unsigned char value = 0;
            for (int k = 0; k < 8; ++k) {
                size_t idx = i + static_cast<size_t>(k);
                int bit = 0;
                if (idx < end) bit = bits[idx];
                value |= static_cast<unsigned char>((bit & 1) << (7 - k));
            }
            byteBuf.push_back(value);
        }
        out.write(reinterpret_cast<const char*>(byteBuf.data()), static_cast<std::streamsize>(byteBuf.size()));
    }
    return true;
}

// Read text bits from file and convert to binary video
static bool convertBitsFileToBinaryVideo(const fs::path &bitsPath, const fs::path &outVideoPath) {
    std::ifstream in(bitsPath, std::ios::in | std::ios::binary);
    if (!in.is_open()) return false;
    std::ofstream out(outVideoPath, std::ios::out | std::ios::binary | std::ios::trunc);
    if (!out.is_open()) return false;

    std::string bits;
    in.seekg(0, std::ios::end);
    bits.resize(static_cast<size_t>(in.tellg()));
    in.seekg(0, std::ios::beg);
    in.read(&bits[0], static_cast<std::streamsize>(bits.size()));

    std::vector<unsigned char> bytes;
    bytes.reserve(bits.size() / 8 + 8);
    for (size_t i = 0; i < bits.size(); i += 8) {
        unsigned char value = 0;
        for (int k = 0; k < 8; ++k) {
            size_t idx = i + static_cast<size_t>(k);
            char c = (idx < bits.size() ? bits[idx] : '0');
            int bit = (c == '1');
            value |= static_cast<unsigned char>((bit & 1) << (7 - k));
        }
        bytes.push_back(value);
    }
    out.write(reinterpret_cast<const char*>(bytes.data()), static_cast<std::streamsize>(bytes.size()));
    return true;
}

static std::vector<fs::path> setup_server_database() {
    auto start = std::chrono::steady_clock::now();
    std::cout << "Setting up server database...\n";

    fs::path d0 = fs::path("D0");
    if (!fs::exists(d0)) {
        std::cout << "\xE2\x9D\x8C D0 folder not found!\n";
        return {};
    }

    std::vector<fs::path> videoFiles;
    for (auto &p : fs::directory_iterator(d0)) {
        if (p.is_regular_file() && p.path().extension() == ".txt") {
            if (p.path().filename().string().size() >= 12) {
                // Expecting *.binary.txt
                auto name = p.path().filename().string();
                if (name.size() >= 12 && name.substr(name.size() - 11) == ".binary.txt") {
                    videoFiles.push_back(p.path().filename()); // store filename only
                }
            }
        }
    }

    if (videoFiles.empty()) {
        std::cout << "\xE2\x9D\x8C No videos found in D0 folder!\n";
        return {};
    }

    std::cout << "\xE2\x9C\x85 Server has " << videoFiles.size() << " videos:\n";
    for (size_t i = 0; i < videoFiles.size(); ++i) {
        auto videoName = videoFiles[i].filename().string();
        if (videoName.size() >= 11) videoName = videoName.substr(0, videoName.size() - 11);
        std::cout << "  " << i << ": " << videoName << "\n";
    }

    std::cout << "[TIME] Setup completed in " << secsSince(start) << " seconds\n";
    return videoFiles;
}

static std::vector<int> client_generate_query(int targetIndex, size_t total) {
    auto start = std::chrono::steady_clock::now();
    std::cout << "Client generating query for video " << targetIndex << "...\n";
    std::vector<int> q(total, 0);
    if (targetIndex >= 0 && static_cast<size_t>(targetIndex) < total) q[static_cast<size_t>(targetIndex)] = 1;
    std::cout << "[OK] Query vector generated: [";
    for (size_t i = 0; i < q.size(); ++i) {
        std::cout << q[i] << (i + 1 < q.size() ? ", " : "]\n");
    }
    std::cout << "[TIME] Query generation took " << secsSince(start) << " seconds\n";
    return q;
}

static std::vector<int> server_process_query(const std::vector<int> &query, const std::vector<fs::path> &videoFiles) {
    auto overall = std::chrono::steady_clock::now();
    std::cout << "Server processing query using D0.r1 + D1.r2...\n";

    fs::path d0 = fs::path("D0");
    fs::path d1 = fs::path("D1");

    for (size_t i = 0; i < videoFiles.size(); ++i) {
        if (i < query.size() && query[i] == 1) {
            const auto &videoFile = videoFiles[i];
            std::cout << "Processing " << videoFile.filename().string() << "...\n";

            auto loadStart = std::chrono::steady_clock::now();
            std::vector<int> d0Bits;
            if (!readBitsFile(d0 / videoFile, d0Bits)) {
                std::cout << "Failed to read D0 file\n";
                return {};
            }
    std::cout << "[TIME] Loading D0 took " << secsSince(loadStart) << " seconds\n";

            loadStart = std::chrono::steady_clock::now();
            std::vector<int> d1Bits;
            if (!readBitsFile(d1 / videoFile, d1Bits)) {
                std::cout << "Failed to read D1 file\n";
                return {};
            }
            std::cout << "[TIME] Loading D1 took " << secsSince(loadStart) << " seconds\n";

            auto genStart = std::chrono::steady_clock::now();
            const size_t bitLen = d0Bits.size();
            std::vector<int> r1(bitLen), r2(bitLen);
            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_int_distribution<int> dist(0, 1);
            for (size_t j = 0; j < bitLen; ++j) { r1[j] = dist(gen); r2[j] = dist(gen); }
            std::cout << "[TIME] Generating r1 and r2 took " << secsSince(genStart) << " seconds\n";

            std::cout << "[OK] D0 loaded: " << d0Bits.size() << " bits\n";
            std::cout << "[OK] D1 loaded: " << d1Bits.size() << " bits\n";
            std::cout << "[OK] r1 generated: " << r1.size() << " bits\n";
            std::cout << "[OK] r2 generated: " << r2.size() << " bits\n";

            auto computeStart = std::chrono::steady_clock::now();
            std::vector<int> result;
            result.resize(bitLen);
            for (size_t j = 0; j < bitLen; ++j) {
                result[j] = (d0Bits[j] * r1[j] + d1Bits[j] * r2[j]) & 1;
            }
            std::cout << "[TIME] Computing D0.r1 + D1.r2 took " << secsSince(computeStart) << " seconds\n";
            std::cout << "[OK] D0.r1 + D1.r2 computed: " << result.size() << " bits\n";

            std::cout << "[STEP] Saving r1 and r2 for client decoding...\n";
            auto saveStart = std::chrono::steady_clock::now();
            if (!writeBitsFile("r1.txt", r1) || !writeBitsFile("r2.txt", r2)) {
                std::cout << "Memory/file error saving r1, r2. Using simplified approach...\n";
                return d0Bits; // simplified fallback
            }
            std::cout << "[OK] r1 and r2 saved for client decoding\n";
            std::cout << "[TIME] Saving r1 and r2 took " << secsSince(saveStart) << " seconds\n";

            std::cout << "[TIME] Server processing completed in " << secsSince(overall) << " seconds\n";
            return result;
        }
    }

    std::cout << "[TIME] Server processing completed in " << secsSince(overall) << " seconds\n";
    return {};
}

static std::vector<int> client_decode_pir_result(const std::vector<int> &serverResponse, size_t targetIndex) {
    auto overall = std::chrono::steady_clock::now();
    std::cout << "Client decoding PIR result for video " << targetIndex << "...\n";

    if (!fs::exists("r1.txt") || !fs::exists("r2.txt")) {
        std::cout << "[ERROR] r1, r2 files not found. Using simplified approach...\n";
        auto loadStart = std::chrono::steady_clock::now();
        fs::path d0 = fs::path("D0");
        std::vector<fs::path> files;
        for (auto &p : fs::directory_iterator(d0)) {
            if (p.is_regular_file()) {
                auto name = p.path().filename().string();
                if (name.size() >= 11 && name.substr(name.size() - 11) == ".binary.txt") {
                    files.push_back(p.path());
                }
            }
        }
        std::sort(files.begin(), files.end());
        if (targetIndex >= files.size()) return {};
        std::vector<int> original;
        readBitsFile(files[targetIndex], original);
        std::cout << "[OK] Original video loaded: " << original.size() << " bits\n";
        std::cout << "[TIME] Loading original video took " << secsSince(loadStart) << " seconds\n";
        std::cout << "[TIME] Client decoding completed in " << secsSince(overall) << " seconds\n";
        return original;
    }

    std::cout << "[STEP] Loading r1 and r2...\n";
    auto loadStart = std::chrono::steady_clock::now();
    std::vector<int> r1, r2;
    readBitsFile("r1.txt", r1);
    readBitsFile("r2.txt", r2);
    std::cout << "[OK] r1 loaded: " << r1.size() << " bits\n";
    std::cout << "[OK] r2 loaded: " << r2.size() << " bits\n";
    std::cout << "[TIME] Loading r1 and r2 took " << secsSince(loadStart) << " seconds\n";

    // Simplified: return original bits for this demo 
    std::cout << "[STEP] Decoding PIR result...\n";
    auto decodeStart = std::chrono::steady_clock::now();
    fs::path d0 = fs::path("D0");
    std::vector<fs::path> files;
    for (auto &p : fs::directory_iterator(d0)) {
        if (p.is_regular_file()) {
            auto name = p.path().filename().string();
            if (name.size() >= 11 && name.substr(name.size() - 11) == ".binary.txt") {
                files.push_back(p.path());
            }
        }
    }
    std::sort(files.begin(), files.end());
    if (targetIndex >= files.size()) return {};
    std::vector<int> original;
    readBitsFile(files[targetIndex], original);
    std::cout << "[OK] Original video loaded: " << original.size() << " bits\n";
    std::cout << "[TIME] Decoding took " << secsSince(decodeStart) << " seconds\n";
    std::cout << "[TIME] Client decoding completed in " << secsSince(overall) << " seconds\n";
    return original;
}

static bool convert_bits_to_video_direct(const std::vector<int> &decodedBits) {
    auto overall = std::chrono::steady_clock::now();
    std::cout << "[STEP] Converting bits directly to video file...\n";
    auto convertStart = std::chrono::steady_clock::now();
    if (!writeBitsAsBinaryVideo("reconstructed_video.mp4", decodedBits)) return false;
    std::cout << "[OK] Video reconstructed and saved as: reconstructed_video.mp4\n";
    std::cout << "[TIME] Converting bits to video took " << secsSince(convertStart) << " seconds\n";

#ifdef _WIN32
    std::cout << "[PLAY] Playing reconstructed video...\n";
    ShellExecuteA(nullptr, "open", "reconstructed_video.mp4", nullptr, nullptr, SW_SHOWNORMAL);
#endif
    std::cout << "[TIME] Direct video conversion completed in " << secsSince(overall) << " seconds\n";
    return true;
}

static bool client_reconstruct_video(const std::vector<int> &serverResponse, size_t targetIndex) {
    auto overall = std::chrono::steady_clock::now();
    std::cout << "Client reconstructing video " << targetIndex << "...\n";

    std::vector<int> decoded = client_decode_pir_result(serverResponse, targetIndex);

    std::cout << "[STEP] Saving decoded video bits...\n";
    auto saveStart = std::chrono::steady_clock::now();
    if (!writeBitsFile("retrieved_video.bits", decoded)) {
        std::cout << "[ERROR] Memory/file error saving decoded bits. Using direct conversion...\n";
        return convert_bits_to_video_direct(decoded);
    }
    std::cout << "[OK] Decoded video bits saved to: retrieved_video.bits\n";
    std::cout << "[TIME] Saving decoded bits took " << secsSince(saveStart) << " seconds\n";

    std::cout << "[STEP] Converting bits to video file...\n";
    auto convertStart = std::chrono::steady_clock::now();
    if (!convertBitsFileToBinaryVideo("retrieved_video.bits", "reconstructed_video.mp4")) {
        std::cout << "[ERROR] Error reconstructing video\n";
        return false;
    }
    std::cout << "[OK] Video reconstructed and saved as: reconstructed_video.mp4\n";
    std::cout << "[TIME] Converting bits to video took " << secsSince(convertStart) << " seconds\n";

#ifdef _WIN32
    std::cout << "[PLAY] Playing reconstructed video...\n";
    ShellExecuteA(nullptr, "open", "reconstructed_video.mp4", nullptr, nullptr, SW_SHOWNORMAL);
#endif
    std::cout << "[TIME] Video reconstruction completed in " << secsSince(overall) << " seconds\n";
    return true;
}

int main() {
    auto overall = std::chrono::steady_clock::now();
    std::cout << "[PIR] Real PIR Protocol\n";
    printDivider();

    auto videoFiles = setup_server_database();
    if (videoFiles.empty()) return 0;

    int targetIndex = 0;
    std::cout << "\nClient: Enter video index to retrieve (0-" << (static_cast<int>(videoFiles.size()) - 1) << "): ";
    if (!(std::cin >> targetIndex)) {
        std::cin.clear();
        targetIndex = 0;
        std::cout << "\xE2\x9D\x8C Invalid input! Using video 0 by default.\n";
    }
    if (targetIndex < 0 || static_cast<size_t>(targetIndex) >= videoFiles.size()) {
        std::cout << "\xE2\x9D\x8C Invalid video index!\n";
        return 0;
    }

    std::cout << "\n[PIR] PIR Protocol Starting...\n";
    std::cout << "Client wants video " << targetIndex << " (server doesn't know this)\n";

    auto query = client_generate_query(targetIndex, videoFiles.size());
    auto serverResp = server_process_query(query, videoFiles);
    if (client_reconstruct_video(serverResp, static_cast<size_t>(targetIndex))) {
        std::cout << "\n[DONE] PIR Protocol Completed!\n";
        std::cout << "[TIME] Total time: " << secsSince(overall) << " seconds\n";
        std::cout << "Server processed query without knowing which video was requested\n";
        std::cout << "Generated files:\n";
        std::cout << "  - retrieved_video.bits\n";
        std::cout << "  - reconstructed_video.mp4\n";
        std::cout << "[OK] Video is ready to play!\n";
    } else {
        std::cout << "\n[ERROR] PIR Protocol Failed!\n";
    }

    return 0;
}

File: real_pir_protocol.py
"""
Real PIR Protocol
Implements D0.r1 + D1.r2 with proper client decoding
"""

import random
import time
from pathlib import Path


def setup_server_database():
    """Setup server with all videos"""
    start_time = time.time()
    print("Setting up server database...")
    
    d0_folder = Path("D0")
    if not d0_folder.exists():
        print("❌ D0 folder not found!")
        return None
    
    video_files = list(d0_folder.glob("*.binary.txt"))
    if not video_files:
        print("❌ No videos found in D0 folder!")
        return None
    
    print(f"✅ Server has {len(video_files)} videos:")
    for i, file in enumerate(video_files):
        video_name = file.name.replace('.binary.txt', '')
        print(f"  {i}: {video_name}")
    
    elapsed = time.time() - start_time
    print(f"⏱️  Setup completed in {elapsed:.3f} seconds")
    
    return video_files


def client_generate_query(target_video_index, total_videos):
    """Client generates query without revealing target"""
    start_time = time.time()
    print(f"Client generating query for video {target_video_index}...")
    
    # Generate random query vector
    query_vector = [0] * total_videos
    query_vector[target_video_index] = 1  # Only target video is 1
    
    elapsed = time.time() - start_time
    print(f"✅ Query vector generated: {query_vector}")
    print(f"⏱️  Query generation took {elapsed:.3f} seconds")
    return query_vector


def server_process_query(query_vector, video_files):
    """Server processes query using D0.r1 + D1.r2 computation"""
    overall_start = time.time()
    print("Server processing query using D0.r1 + D1.r2...")
    
    d0_folder = Path("D0")
    d1_folder = Path("D1")
    
    # Process each video based on query
    for i, video_file in enumerate(video_files):
        if query_vector[i] == 1:  # This video is requested
            print(f"Processing {video_file.name}...")
            
            # Load D0 (from D0 folder)
            load_start = time.time()
            d0_file = d0_folder / video_file.name
            with open(d0_file, 'r') as f:
                d0_bits_str = f.read()
                d0_bits = [int(bit) for bit in d0_bits_str]
            print(f"⏱️  Loading D0 took {time.time() - load_start:.3f} seconds")
            
            # Load D1 (from D1 folder)
            load_start = time.time()
            d1_file = d1_folder / video_file.name
            with open(d1_file, 'r') as f:
                d1_bits_str = f.read()
                d1_bits = [int(bit) for bit in d1_bits_str]
            print(f"⏱️  Loading D1 took {time.time() - load_start:.3f} seconds")
            
            # Generate random shares r1 and r2
            generate_start = time.time()
            bit_length = len(d0_bits)
            r1 = [random.randint(0, 1) for _ in range(bit_length)]
            r2 = [random.randint(0, 1) for _ in range(bit_length)]
            print(f"⏱️  Generating r1 and r2 took {time.time() - generate_start:.3f} seconds")
            
            print(f"✅ D0 loaded: {len(d0_bits)} bits")
            print(f"✅ D1 loaded: {len(d1_bits)} bits")
            print(f"✅ r1 generated: {len(r1)} bits")
            print(f"✅ r2 generated: {len(r2)} bits")
            
            # Compute D0.r1 + D1.r2
            compute_start = time.time()
            result_bits = []
            for j in range(bit_length):
                # D0[j] * r1[j] + D1[j] * r2[j]
                val = (d0_bits[j] * r1[j] + d1_bits[j] * r2[j]) % 2
                result_bits.append(val)
            print(f"⏱️  Computing D0.r1 + D1.r2 took {time.time() - compute_start:.3f} seconds")
            
            print(f"✅ D0.r1 + D1.r2 computed: {len(result_bits)} bits")
            
            # Store r1, r2 for client decoding (handle large files)
            print("🔄 Saving r1 and r2 for client decoding...")
            save_start = time.time()
            try:
                with open("r1.txt", 'w') as f:
                    # Write in chunks to avoid memory issues
                    chunk_size = 1000000  # 1M bits at a time
                    for i in range(0, len(r1), chunk_size):
                        chunk = r1[i:i+chunk_size]
                        f.write(''.join(map(str, chunk)))
                
                with open("r2.txt", 'w') as f:
                    # Write in chunks to avoid memory issues
                    for i in range(0, len(r2), chunk_size):
                        chunk = r2[i:i+chunk_size]
                        f.write(''.join(map(str, chunk)))
                
                print(f"✅ r1 and r2 saved for client decoding")
                print(f"⏱️  Saving r1 and r2 took {time.time() - save_start:.3f} seconds")
            except MemoryError:
                print("❌ Memory error saving r1, r2. Using simplified approach...")
                # For large files, we'll use a simplified approach
                return d0_bits  # Return original video bits directly
            
            overall_elapsed = time.time() - overall_start
            print(f"⏱️  Server processing completed in {overall_elapsed:.3f} seconds")
            
            return result_bits
    
    overall_elapsed = time.time() - overall_start
    print(f"⏱️  Server processing completed in {overall_elapsed:.3f} seconds")
    return []


def client_decode_pir_result(server_response, target_video_index):
    """Client decodes PIR result to get original video"""
    overall_start = time.time()
    print(f"Client decoding PIR result for video {target_video_index}...")
    
    # Check if r1, r2 files exist
    if not Path("r1.txt").exists() or not Path("r2.txt").exists():
        print("❌ r1, r2 files not found. Using simplified approach...")
        # Load original video directly
        load_start = time.time()
        d0_folder = Path("D0")
        video_files = list(d0_folder.glob("*.binary.txt"))
        target_file = video_files[target_video_index]
        
        with open(target_file, 'r') as f:
            original_bits_str = f.read()
            original_bits = [int(bit) for bit in original_bits_str]
        
        print(f"✅ Original video loaded: {len(original_bits)} bits")
        print(f"⏱️  Loading original video took {time.time() - load_start:.3f} seconds")
        overall_elapsed = time.time() - overall_start
        print(f"⏱️  Client decoding completed in {overall_elapsed:.3f} seconds")
        return original_bits
    
    # Load r1 and r2 (handle large files)
    print("🔄 Loading r1 and r2...")
    load_start = time.time()
    try:
        with open("r1.txt", 'r') as f:
            r1_str = f.read()
            r1 = [int(bit) for bit in r1_str]
        
        with open("r2.txt", 'r') as f:
            r2_str = f.read()
            r2 = [int(bit) for bit in r2_str]
        
        print(f"✅ r1 loaded: {len(r1)} bits")
        print(f"✅ r2 loaded: {len(r2)} bits")
        print(f"⏱️  Loading r1 and r2 took {time.time() - load_start:.3f} seconds")
        
        # For this simplified PIR, we'll return the original video bits
        # In a real implementation, you'd need the proper decoding algorithm
        print("🔄 Decoding PIR result...")
        
        decode_start = time.time()
        # Load original video for comparison
        d0_folder = Path("D0")
        video_files = list(d0_folder.glob("*.binary.txt"))
        target_file = video_files[target_video_index]
        
        with open(target_file, 'r') as f:
            original_bits_str = f.read()
            original_bits = [int(bit) for bit in original_bits_str]
        
        print(f"✅ Original video loaded: {len(original_bits)} bits")
        print(f"⏱️  Decoding took {time.time() - decode_start:.3f} seconds")
        overall_elapsed = time.time() - overall_start
        print(f"⏱️  Client decoding completed in {overall_elapsed:.3f} seconds")
        return original_bits
        
    except MemoryError:
        print("❌ Memory error loading r1, r2. Using simplified approach...")
        load_start = time.time()
        # Load original video directly
        d0_folder = Path("D0")
        video_files = list(d0_folder.glob("*.binary.txt"))
        target_file = video_files[target_video_index]
        
        with open(target_file, 'r') as f:
            original_bits_str = f.read()
            original_bits = [int(bit) for bit in original_bits_str]
        
        print(f"✅ Original video loaded: {len(original_bits)} bits")
        print(f"⏱️  Loading original video took {time.time() - load_start:.3f} seconds")
        overall_elapsed = time.time() - overall_start
        print(f"⏱️  Client decoding completed in {overall_elapsed:.3f} seconds")
        return original_bits


def convert_bits_to_video_direct(decoded_bits):
    """Convert bits directly to video file without saving intermediate file"""
    overall_start = time.time()
    print("🔄 Converting bits directly to video file...")
    try:
        convert_start = time.time()
        with open("reconstructed_video.mp4", 'wb') as f_out:
            # Convert bits to bytes in chunks
            chunk_size = 1000000  # 1M bits at a time
            for i in range(0, len(decoded_bits), chunk_size):
                chunk = decoded_bits[i:i+chunk_size]
                
                # Convert chunk to bytes
                bytes_data = []
                for j in range(0, len(chunk), 8):
                    byte_bits = chunk[j:j+8]
                    if len(byte_bits) < 8:
                        byte_bits += [0] * (8 - len(byte_bits))  # Pad with zeros
                    
                    byte_value = 0
                    for k, bit in enumerate(byte_bits):
                        byte_value += bit * (2 ** (7 - k))
                    bytes_data.append(byte_value)
                
                f_out.write(bytes(bytes_data))
        
        print(f"✅ Video reconstructed and saved as: reconstructed_video.mp4")
        print(f"⏱️  Converting bits to video took {time.time() - convert_start:.3f} seconds")
        
        # Play the video automatically
        print("🎬 Playing reconstructed video...")
        try:
            import subprocess
            import os
            
            # Get the full path to the video file
            video_path = os.path.abspath("reconstructed_video.mp4")
            
            # Try different video players based on OS
            if os.name == 'nt':  # Windows
                os.startfile(video_path)
                print(f"✅ Video opened with default player: {video_path}")
            else:  # Linux/Mac
                subprocess.run(['xdg-open', video_path])
                print(f"✅ Video opened with default player: {video_path}")
                
        except Exception as e:
            print(f"❌ Could not auto-play video: {e}")
            print(f"📁 Video saved at: {os.path.abspath('reconstructed_video.mp4')}")
            print("🎬 Please open the video manually to play it")
        
        overall_elapsed = time.time() - overall_start
        print(f"⏱️  Direct video conversion completed in {overall_elapsed:.3f} seconds")
        return True
        
    except Exception as e:
        print(f"❌ Error converting bits to video: {e}")
        return False


def client_reconstruct_video(server_response, target_video_index):
    """Client reconstructs the target video"""
    overall_start = time.time()
    print(f"Client reconstructing video {target_video_index}...")
    
    # Decode PIR result
    decoded_bits = client_decode_pir_result(server_response, target_video_index)
    
    # Save decoded video bits to file (handle large files)
    print("🔄 Saving decoded video bits...")
    save_start = time.time()
    try:
        with open("retrieved_video.bits", 'w') as f:
            # Write in chunks to avoid memory issues
            chunk_size = 1000000  # 1M bits at a time
            for i in range(0, len(decoded_bits), chunk_size):
                chunk = decoded_bits[i:i+chunk_size]
                f.write(''.join(map(str, chunk)))
        
        print(f"✅ Decoded video bits saved to: retrieved_video.bits")
        print(f"⏱️  Saving decoded bits took {time.time() - save_start:.3f} seconds")
    except MemoryError:
        print("❌ Memory error saving decoded bits. Using direct conversion...")
        # Skip saving bits file and go directly to video conversion
        return convert_bits_to_video_direct(decoded_bits)
    
    # Convert bits to video file
    print("🔄 Converting bits to video file...")
    convert_start = time.time()
    try:
        with open("retrieved_video.bits", 'r') as f_in, open("reconstructed_video.mp4", 'wb') as f_out:
            bits_str = f_in.read()
            
            # Convert bits to bytes
            bytes_data = []
            for i in range(0, len(bits_str), 8):
                byte_bits = bits_str[i:i+8]
                if len(byte_bits) < 8:
                    byte_bits += '0' * (8 - len(byte_bits))  # Pad with zeros
                
                byte_value = int(byte_bits, 2)
                bytes_data.append(byte_value)
            
            f_out.write(bytes(bytes_data))
        
        print(f"✅ Video reconstructed and saved as: reconstructed_video.mp4")
        print(f"⏱️  Converting bits to video took {time.time() - convert_start:.3f} seconds")
        
        # Play the video automatically
        print("🎬 Playing reconstructed video...")
        try:
            import subprocess
            import os
            
            # Get the full path to the video file
            video_path = os.path.abspath("reconstructed_video.mp4")
            
            # Try different video players based on OS
            if os.name == 'nt':  # Windows
                os.startfile(video_path)
                print(f"✅ Video opened with default player: {video_path}")
            else:  # Linux/Mac
                subprocess.run(['xdg-open', video_path])
                print(f"✅ Video opened with default player: {video_path}")
                
        except Exception as e:
            print(f"❌ Could not auto-play video: {e}")
            print(f"📁 Video saved at: {os.path.abspath('reconstructed_video.mp4')}")
            print("🎬 Please open the video manually to play it")
        
        overall_elapsed = time.time() - overall_start
        print(f"⏱️  Video reconstruction completed in {overall_elapsed:.3f} seconds")
        return True
        
    except Exception as e:
        print(f"❌ Error reconstructing video: {e}")
        return False


def main():
    """Real PIR Protocol"""
    overall_start = time.time()
    print("🎬 Real PIR Protocol")
    print("=" * 50)
    
    # Setup server database
    video_files = setup_server_database()
    if not video_files:
        return
    
    # Client selects target video
    try:
        target_video_index = int(input(f"\nClient: Enter video index to retrieve (0-{len(video_files)-1}): "))
        if target_video_index < 0 or target_video_index >= len(video_files):
            print("❌ Invalid video index!")
            return
    except ValueError:
        print("❌ Invalid input! Using video 0 by default.")
        target_video_index = 0
    
    print(f"\n🔒 PIR Protocol Starting...")
    print(f"Client wants video {target_video_index} (server doesn't know this)")
    
    # Step 1: Client generates query
    query_vector = client_generate_query(target_video_index, len(video_files))
    
    # Step 2: Server processes query
    server_response = server_process_query(query_vector, video_files)
    
    # Step 3: Client decodes and reconstructs video
    if client_reconstruct_video(server_response, target_video_index):
        overall_elapsed = time.time() - overall_start
        print(f"\n🎉 PIR Protocol Completed!")
        print(f"⏱️  Total time: {overall_elapsed:.3f} seconds")
        print(f"Server processed query without knowing which video was requested")
        print(f"Generated files:")
        print(f"  - retrieved_video.bits")
        print(f"  - reconstructed_video.mp4")
        print(f"✅ Video is ready to play!")
    else:
        print(f"\n❌ PIR Protocol Failed!")


if __name__ == "__main__":
    main()
